test/node/object_id_tests.js

```diff
@@ -24,6 +25,195 @@ describe('ObjectId', function () {
+    };
+    expect(() => new ObjectId(objectBufferId).toHexString()).to.not.throw(TypeError);
+    expect(Buffer.from(new ObjectId(objectBufferId).id).equals(bufferToHex));
+
+    expect(() => new ObjectId(objectBufferFromArray).toHexString()).to.not.throw(TypeError);
+    expect(Buffer.from(new ObjectId(objectBufferFromArray).id).equals(bufferToHex));
+  });
+
+  it('should throw error if object with non-Buffer non-string id is passed in', function () {
+    var objectNumId = {
+      id: 5
+    };
+    var objectNullId = {
+      id: null
+    };
+    expect(() => new ObjectId(objectNumId)).to.throw(TypeError);
+    expect(() => new ObjectId(objectNullId)).to.throw(TypeError);
+  });
+
+  it('should throw an error if object with invalid string id is passed in', function () {
+    var objectInvalidString = {
+      id: 'FFFFFFFFFFFFFFFFFFFFFFFG'
+    };
+    expect(() => new ObjectId(objectInvalidString)).to.throw(TypeError);
+  });
+
+  it('should correctly create ObjectId from object with invalid string id and toHexString method', function () {
+    function newToHexString() {
+      return 'BBBBBBBBBBBBBBBBBBBBBBBB';
+    }
+    var objectInvalidString = {
+      id: 'FFFFFFFFFFFFFFFFFFFFFFFG',
+      toHexString: newToHexString
+    };
+    var buf = Buffer.from('BBBBBBBBBBBBBBBBBBBBBBBB', 'hex');
+    expect(() => new ObjectId(objectInvalidString)).to.not.throw(TypeError);
+    expect(Buffer.from(new ObjectId(objectInvalidString).id).equals(buf));
+  });
+
+  it('should throw an error if object with invalid Buffer id is passed in', function () {
+    var objectInvalidBuffer = {
+      id: Buffer.from([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13])
+    };
+    expect(() => new ObjectId(objectInvalidBuffer)).to.throw(TypeError);
+  });
+
+  it('should correctly create ObjectId from object with invalid Buffer id and toHexString method', function () {
+    function newToHexString() {
+      return 'BBBBBBBBBBBBBBBBBBBBBBBB';
+    }
+    var objectInvalidBuffer = {
+      id: Buffer.from([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]),
+      toHexString: newToHexString
+    };
+    var buf = Buffer.from('BBBBBBBBBBBBBBBBBBBBBBBB', 'hex');
+    expect(() => new ObjectId(objectInvalidBuffer)).to.not.throw(TypeError);
+    expect(Buffer.from(new ObjectId(objectInvalidBuffer).id).equals(buf));
+  });
+
+  it('should correctly create ObjectId from object with objectIdLike properties', function () {
+    var tmp = new ObjectId();
+    var objectIdLike = {
+      id: tmp.id,
+      toHexString: function () {
+        return tmp.toHexString();
+      }
+    };
+
+    expect(() => new ObjectId(objectIdLike).toHexString()).to.not.throw(TypeError);
+  });
+
+  it('should correctly create ObjectId from number or null', function () {
+    expect(() => new ObjectId(42).toHexString()).to.not.throw();
+    expect(() => new ObjectId(0x2a).toHexString()).to.not.throw();
+    expect(() => new ObjectId(NaN).toHexString()).to.not.throw();
```

Comment:
Interesting.. NaN doesn't throw, it's probably silently doing something undesirable though, can you add a floating point number here as well.

In the generate() function this will get passed to buffer.writeUInt32LE(), which may do flooring? We should add a test to check.